### PR TITLE
Fix `dist_mapper()` in packagedcode/npm.py

### PR DESCRIPTION
**Pull request by majurg**
* Properly handle base64 encoded strings in the npm dist_mapper()
  function.
* Add test case

Addresses: #2256

Signed-off-by: Steven Esser <sesser@nexb.com>

### CODE REVIEW

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks! See my notes inline wrt the sha512 encoding, as I am not sure this is correct.

File: src/packagedcode/npm.py [hunk 1]
```diff
@@ -577,7 +579,7 @@ def dist_mapper(dist, package):
         algo, _, b64value = integrity.partition('-')
         assert 'sha512' == algo
         algo = algo.lower()
-        sha512 = b64value.decode('base64').encode('hex')
+        sha512 = hashlib.sha512(base64.b64decode(b64value)).hexdigest()
```
[the remark this answers is not in the document]
Comment (majurg) [reply src/packagedcode/npm.py:582]: Yes I was thinking about this last night and realized all I have to do is b64decode the value to get the proper `sha512` (and not rehashing)

File: src/packagedcode/npm.py [hunk 2]
```diff
@@ -577,7 +579,7 @@ def dist_mapper(dist, package):
         algo, _, b64value = integrity.partition('-')
         assert 'sha512' == algo
         algo = algo.lower()
```
[the remark this answers is not in the document]
Comment (majurg) [reply src/packagedcode/npm.py:581]: good point, I will make this change

File: src/packagedcode/npm.py [hunk 3]
```diff
@@ -26,9 +26,11 @@
 from __future__ import print_function
 from __future__ import unicode_literals
 
+import base64
 from collections import defaultdict
 from collections import OrderedDict
 from functools import partial
+import hashlib
```
Comment (pombredanne) [inline src/packagedcode/npm.py:33]: Do you still need hashlib now?

Comment (majurg) [reply src/packagedcode/npm.py]: Nope, the latest commit removes this.

File: src/packagedcode/npm.py [hunk 4]
```diff
@@ -575,10 +576,16 @@ def dist_mapper(dist, package):
...
         algo, _, b64value = integrity.partition('-')
-        assert 'sha512' == algo
         algo = algo.lower()
-        sha512 = b64value.decode('base64').encode('hex')
+        assert 'sha512' == algo
+
+        decoded_b64value = base64.b64decode(b64value)
+        if type(decoded_b64value) == str:
+            sha512 = decoded_b64value.encode('hex')
+        elif type(decoded_b64value) == bytes:
```
Comment (pombredanne) [inline src/packagedcode/npm.py:585]: That may not work between Py2 and Py3.
You may instead try this:
```
        if isinstance(decoded_b64value, string_types):
            sha512 = decoded_b64value.encode('hex')
        elif isinstance(decoded_b64value, binary_type):
....
```

Comment (pombredanne) [reply src/packagedcode/npm.py:585]: Note that I think that b64decode results may always be bytes on Python 2 or Python 3 so you may not need the type test

Comment (majurg) [reply src/packagedcode/npm.py:585]: I made this change in response to test failures on the python2 side. The latest code now works on both py3 and py2. `b64decode` returns bytes on python3 and string in python2, hence my code above. 

Unless I am missing something, why would `isinstance` be preferable here?

Comment (pombredanne) [reply src/packagedcode/npm.py:585]: > Unless I am missing something, why would isinstance be preferable here?

that's catching subclasses too and I use it as an automatism but that's not the main point. Rather using `six.string_types` and `six.binary_types` is what matters more IMHO for portability.

**Review by pombredanne — APPROVED** [review]
Thanks for all the fixes. LGTM! :+1:

**Comment by pombredanne** [comment]
Merging now that we are all green. :green_apple:  Thank you ++

**Pull request merged**